doc/doxygen/src/driver-guide.md

```diff
@@ -266,18 +266,30 @@ int16_t tmpabc_read(const tmpabc_t *dev)
 }
 @endcode
 
-Whenever you implement status/error return values in your driver, they
-should be named, meaning that the driver MUST define an enum assigning names to
-the actual used value, e.g.
+Whenever you implement status/error return values in your driver, magic numbers
+must ***not*** be used. Instead, use negative `errno` codes (such as `-EIO` for
+an input/output error) and *document* every single error code that can be
+return for each function using the `reval` Doxygen command. E.g. like this:
 
-@code{.c}
-enum {
-    TMPABC_OK    = 0,       /**< all went as expected */
-    TMPABC_NOI2C = -1,      /**< error using the I2C bus */
-    TMPABC_NODEV = -2       /**< no device with the configured address found on the bus */
-};
+@code
+/**
+ * @brief               Initialize the foo device and its descriptor
+ * @param[out]  dev     Device descriptor to initialized
+ * @param[in]   params  Device initialization parameters
+ *
+ * @retval      0       Success
+ * @retval      -ENXIO  No foo device connected to the I2C bus
+ * @retval      -ENODEV Device using the foo I2C address is not a foo device
+ * etc.
+ */
+int foo_init(foo_t *dev, const foo_params_t *params);
 @endcode
 
+You can multiplex this information whenever this is reasonable, as long as a
+negative return value indicates an error without exceptions. E.g. when reading
+all values from a sensor's buffer the number of successfully read entries could
+be returned.
```

Comment:
> E.g. when reading ￼all values from a sensor's buffer the number of successfully read entries could ￼be returned.

I don't think that's a good example.
When you read a 4 byte sensor value or an 8-byte address, returning the size and having the user to check if the size matches the expectation is more cumbersome than just returning 0 or error.

What is the user supposed to do if the size would not match the expectation? Wouldn't it be much better to return `-EIO` instead, especially when there is no way to resume reading at an offset? (And the API user shouldn't have to deal with that, either the driver or a intermediate layer (e.g. MTD) would handle chunked reads / writes)

Reply:
This is not an ideal example. E.g. for what this multiplexing makes sense would be the `recv()` function of a network driver, where you don't have any expectations about the size.

But let's says the sensor is a Photoplethysmogroph (optical heart rate sensor). Those are under the hood all ADCs that sample the voltage drop of a photo resistor that is on the one side of your finger, and a red and IR led is on the other. The red and IR Leds flash in turns, so that every second sample depends on the red light intensity shining through your finger, and every other on the IR light intensity. When plotted, those two curves will look a bit like a sine wave. And the frequency of this waves is the heart rate.

So, one could add a buffer to those ADCs that could hold up e.g. 64 samples. If you read the data, you would get on success 0 to 64 samples, depending on how many new samples have been measured since the last read. In this scenario IMO the API for reading should expect a buffer for 64 sample and the read function should return the number of samples actually read, or a negative error on failure.

(This buffer would in low power scenarios in which long term heart rate monitoring is needed indeed sense, as this would reduce the number of wake ups of the MCU.)

I could just drop the example. I think it will be obvious to the programmer when his API is a good case for multiplexing error codes with useful information.

Reply:
I dropped the example. An example that needs explaining is certainly not a good one.